egglog-bridge/src/lib.rs

```diff
@@ -84,6 +85,49 @@ pub struct EGraph {
     /// as a proof object with a given number of parameters is added.
     reason_tables: IndexMap<usize /* arity */, TableId>,
     term_tables: IndexMap<usize /* arity */, TableId>,
+    /// Union-find table that records the stable term id for each provisional term id.
+    ///
+    /// This table is only used when proofs are enabled. `core-relations` has a concept of a
+    /// "predicted value" to handle lookups for rows on tables that haven't been created yet. For
+    /// example, in the associativity rewrite:
+    ///
+    /// > (rewrite (add (add x y) z) (add x (add y z)))
+    ///
+    /// The RHS of the rule will check if `(add y z)` is present in the database, and if it isn't
+    /// it will create a new id and insert `(add y z) => id`. This `id` is cached so subsequent
+    /// uses of `(add y z)` will still see `id` within the same rule.
+    ///
+    /// This cache is _local_: so different threads can potentially see different values for `id`.
+    /// This is fine as all provisional values for `id` will get unioned together and the database
+    /// will be consistent after congruence closure runs. This same logic, however, does not hold
+    /// true for proofs.
+    ///
+    /// When proofs mint a new `id` for a row it is used both as the canonical e-class id in the main
+    /// e-graph and also the *term* id for that specific row. Term ids are used to reconstruct
+    /// proofs and never change. e-class ids can change depending on the result of a UF `union`
+    /// operation. For example, consider these three terms:
+    ///
+    /// > x => term: id0, canon: id0
+    /// > (add x 0) => term: id1, canon: id1
+    /// > (add 0 x) => term: id2, canon: id2
+    ///
+    /// If we have rules like `(add x 0) => x` and `(add x y) => (add y x)` then our e-graph
+    /// may look like this:
+    ///
+    /// > x => term: id0, canon: id0
+    /// > (add x 0) => term: id1, canon: id0
+    /// > (add 0 x) => term: id2, canon: id0
+    ///
+    /// So in other words, while canonical ids change over time and need not be unique to a term,
+    /// term ids are specific to the specific shape of the (head of the) term when it was first
+    /// instantiated. This is a problem for local caching, because it means that we can have
+    /// multiple term values for the same row. How do we pick the real one?
+    ///
+    /// We have a separate union-find. Duplicate insertions into the term table cause `union`s on
+    /// this union-find, and future lookups of this table canonicalize ids with respect to this
+    /// union-find. In this way, it's a subset of the full union-find in the `uf_table` row, only
+    /// used to resolved temporary inconsistencies in cached term values.
+    term_consistency_table: TableId,
```

Comment:
Ok, this makes sense to me and I don't know how you would do it otherwise. Parallel egglog sure makes things complicated! Afaict this is only necessary in parallel mode, right? It's fine to keep it on regardless, but could consider disabling in sequential mode.

Reply:
Oh I see, okay thanks